lib/backends/smartos/tasks/recovery_config.js

```diff
@@ -0,0 +1,282 @@
+                    return;
+                }
+                self.progress(50);
+                cb();
+            });
+        },
+
+        function doAction(cb) {
+            var cmd = '/usr/sbin/kbmadm';
+            var args = ['recovery'];
+            if (action === 'activate') {
+                args.push('activate', zpool);
+            } else {
+                args.push('add', '-t', templateFile, '-r', token, zpool);
+            }
+
+            self.log.debug({cmdline: cmd + ' ' + args.join(' ')}, 'Executing');
+            execFile(cmd, args, function (err, stdout, stderr) {
+                if (err) {
+                    self.log.error({
+                        err: err
+                    }, util.format('Failed to % recovery configuration',
+                        action));
+                    cb(err);
+                    return;
+                }
+                self.log.debug({
+                    stdout: stdout,
+                    stderr: stderr
+                }, util.format('% recovery configuration'));
+                self.progress(65);
+                cb();
+            });
+        },
+
+        function getSysinfoUpdated(cb) {
+            execFile('/usr/bin/sysinfo', ['-f'], function (err, stdo, stde) {
+                if (err) {
+                    cb(Error(stde.toString()));
+                    return;
+                }
+                var obj = JSON.parse(stdo.toString());
+                zpoolRecovery = obj['Zpool Recovery'] || zpoolRecovery;
+                self.progress(75);
+                cb();
+            });
+        },
+        /*
+         * Sysinfo will be "automatically" updated into CNAPI at the given
+         * period of time. We want KBMAPI updated too.
+         */
+        function postUpdatesToKbmapi(cb) {
+            var url = 'http://' + kbmapiaddr;
+
+            var restifyOptions = {
+                url: url,
+                connectTimeout: 5000,
+                requestTimeout: 5000
+            };
+
+            logger.info('kbmapi ip was %s', kbmapiaddr);
+            var client = restify.createJsonClient(restifyOptions);
+
+            var params = {
+                zpool_recovery: {},
+                cn_uuid: server_uuid
+            };
+
+            logger.info({zpool_recovery: zpoolRecovery}, 'zpool recovery');
+            Object.keys(zpoolRecovery).forEach(function uuidfromstr(k) {
+                params.zpool_recovery[k] =
+                    repeatableUUIDFromHexString(zpoolRecovery[k]);
+            });
+
+            if (token) {
+                params.recovery_token =
+                    repeatableUUIDFromHexString(sha512str(token));
+            }
+
+            var kbmapi_path = '/pivtokens/' + guid + '/recovery-tokens';
+            client.put(kbmapi_path, params, function kbmapiCb(err) {
+                if (err) {
+                    logger.warn({
+                        error: err
+                    }, 'posting info to kbmapi');
```

Comment:
Should this be returning some type of error in this case?

Reply:
Indeed. forgot to add it to the callback after adding the logging branch above.